cpp/src/arrow/util/vector.h

```diff
@@ -81,5 +84,53 @@ std::vector<T> FilterVector(std::vector<T> values, Predicate&& predicate) {
   return values;
 }
 
+/// \brief Like MapVector, but where the function can fail.
+template <typename Fn, typename From = internal::call_traits::argument_type<0, Fn>,
+          typename To = typename internal::call_traits::return_type<Fn>::ValueType>
+Result<std::vector<To>> MaybeMapVector(Fn&& map, const std::vector<From>& src) {
+  std::vector<To> out;
+  out.reserve(src.size());
+  ARROW_RETURN_NOT_OK(
+      MaybeTransform(src.begin(), src.end(), std::back_inserter(out), map));
+  return out;
+}
+
+template <typename Fn, typename From = internal::call_traits::argument_type<0, Fn>,
+          typename To = typename internal::call_traits::return_type<Fn>>
+std::vector<To> MapVector(Fn&& map, const std::vector<From>& source) {
+  std::vector<To> out;
+  out.reserve(source.size());
+  std::transform(source.begin(), source.end(), std::back_inserter(out), map);
+  return out;
+}
+
+template <typename T>
+std::vector<T> FlattenVectors(const std::vector<std::vector<T>>& vecs) {
+  std::size_t sum = 0;
+  for (const auto& vec : vecs) {
+    sum += vec.size();
+  }
+  std::vector<T> out;
+  out.reserve(sum);
+  for (const auto& vec : vecs) {
+    out.insert(out.end(), vec.begin(), vec.end());
+  }
+  return out;
+}
+
+template <typename T>
+Result<std::vector<T>> UnwrapOrRaise(std::vector<Result<T>> results) {
```

Comment:
I think it's enough to take `std::vector<Result<T>>&& results`. No copy is done of the vector itself, we're just moving its individual elements.

Reply:
Yes, so this saves the move constructor call?  I've changed to this.

Reply:
Ping for @bkietz In case you have any good quick solution

Reply:
Sorry, this comment was added in the wrong spot and Antoine already solved the problem so please ignore this if you haven't already.